pkg/plugins/runtime/k8s/controllers/pod_controller.go

```diff
@@ -382,15 +426,52 @@ func (r *PodReconciler) SetupWithManager(mgr kube_ctrl.Manager, maxConcurrentRec
 		For(&kube_core.Pod{}).
 		// on Service update reconcile affected Pods (all Pods selected by this service)
 		Watches(&kube_core.Service{}, kube_handler.EnqueueRequestsFromMapFunc(ServiceToPodsMapper(r.Log, mgr.GetClient()))).
+		Watches(&kube_discovery.EndpointSlice{}, kube_handler.EnqueueRequestsFromMapFunc(EndpointSliceToPodsMapper(r.Log, mgr.GetClient()))).
 		Complete(r)
 }
 
 func ServiceToPodsMapper(l logr.Logger, client kube_client.Client) kube_handler.MapFunc {
 	l = l.WithName("service-to-pods-mapper")
 	return func(ctx context.Context, obj kube_client.Object) []kube_reconcile.Request {
+		svc := obj.(*kube_core.Service)
+		l := l.WithValues(
+			"Service",
+			kube_types.NamespacedName{Namespace: svc.Namespace, Name: svc.Name},
+		)
 		// List Pods in the same namespace as a Service
 		pods := &kube_core.PodList{}
-		if err := client.List(ctx, pods, kube_client.InNamespace(obj.GetNamespace()), kube_client.MatchingLabels(obj.(*kube_core.Service).Spec.Selector)); err != nil {
+		if len(svc.Spec.Selector) == 0 {
+			endpointSlices := &kube_discovery.EndpointSliceList{}
+			if err := client.List(
+				ctx,
+				endpointSlices,
+				kube_client.InNamespace(svc.Namespace),
+				kube_client.MatchingLabels(map[string]string{
+					kube_discovery.LabelServiceName: svc.Name,
+				}),
+			); err != nil {
+				l.Error(err, "failed to fetch EndpointSlices")
+				return nil
+			}
+			var req []kube_reconcile.Request
+			for _, slice := range endpointSlices.Items {
+				l = l.WithValues(
+					"EndpointSlice",
+					kube_types.NamespacedName{Namespace: slice.Namespace, Name: slice.Name},
+				)
+				for _, endpoint := range slice.Endpoints {
+					l = l.WithValues("Endpoint", endpoint)
+					if endpoint.TargetRef != nil && endpoint.TargetRef.Kind == "Pod" && endpoint.TargetRef.APIVersion == kube_core.SchemeGroupVersion.String() {
+						req = append(req, kube_reconcile.Request{
+							NamespacedName: kube_types.NamespacedName{Name: endpoint.TargetRef.Name, Namespace: endpoint.TargetRef.Namespace},
+						})
+					}
+				}
+			}
+			return req
+		}
+
+		if err := client.List(ctx, pods, kube_client.InNamespace(obj.GetNamespace()), kube_client.MatchingLabels(svc.Spec.Selector)); err != nil {
 			l.WithValues("service", obj.GetName()).Error(err, "failed to fetch Pods")
```

Comment:
this doesn't do anything right? because we never use this new logger right? I know it's not code from this PR but seems just wrong... also it will contain `EndpointSlice` and `Endpoint` from last loop cycles which is not right 🤔

Reply:
fixed the unused loggers but this calls `Error` so it's important